Uploader/TraceFileScanner.cs

```diff
@@ -0,0 +1,147 @@
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+/// <summary>
+/// Scans the trace directory for trace files that are ready to upload or archive.
+/// </summary>
+public class TraceFileScanner
+{
+    private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+    private static readonly Regex TraceFileRegex = new Regex(@"^coverage_\d*_\d*.txt$");
+
+    private readonly string traceDirectory;
+    private readonly Regex versionAssemblyRegex;
+    private readonly IFileSystem fileSystem;
+
+    public TraceFileScanner(string traceDirectory, string versionAssembly, IFileSystem fileSystem)
+    {
+        this.traceDirectory = traceDirectory;
+        this.versionAssemblyRegex = new Regex(@"^Assembly=" + Regex.Escape(versionAssembly) + @".*Version:([^ ]*).*", RegexOptions.IgnoreCase);
+        this.fileSystem = fileSystem;
+    }
+
+    /// <summary>
+    /// Returns all trace files that can be uploaded or archived.
+    /// </summary>
+    public IEnumerable<ScannedFile> ListTraceFilesReadyForUpload()
+    {
+        List<string> files;
+        try
+        {
+            files = fileSystem.Directory.EnumerateFiles(traceDirectory).ToList();
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Unable to list files in {traceDirectory}. Will retry later", traceDirectory);
+            yield break;
+        }
+        
+        logger.Debug("Scanning {fileCount} files", files.Count);
+        foreach (string filePath in files)
+        {
+            string fileName = Path.GetFileName(filePath);
+            if (!IsTraceFile(fileName))
+            {
+                logger.Debug("Skipping file that does not look like a trace file: {unknownFilePath}", filePath);
+                continue;
+            }
+
+            ScannedFile scannedFile = ScanFile(filePath);
+            if (scannedFile != null)
+            {
+                yield return scannedFile;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Scans the given file path and returns the resulting ScannedFile or null in case the file should be ignored.
+    /// </summary>
+    private ScannedFile ScanFile(string filePath)
+    {
+        string[] lines;
+        try
+        {
+            lines = fileSystem.File.ReadAllLines(filePath);
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Unable to read from file {trace}. Ignoring this file", filePath);
+            return null;
+        }
+
+        if (!IsFinished(lines))
+        {
+            logger.Debug("Ignoring unfinished trace {tracePath}", filePath);
+            return null;
+        }
+
+        string version = FindVersion(lines, filePath);
+        return new ScannedFile
+        {
+            FilePath = filePath,
+            Version = version,
+        };
+    }
+
+    private bool IsFinished(string[] lines)
+    {
+        return lines.Any(line => line.StartsWith("Jitted=") || line.StartsWith("Inlined="));
```

Comment:
I think we will remove these entries in a future version (other PR). Either implement a generic end line in C++ OR simply check whether the file is no longer locked here?

Reply:
don't know how to check for file locking (especially considering network shares etc). I think that's brittle. I prefer scanning the file contents. IMO the other pull request should implement this since I have no idea what happened there and don't want to speculate here

Reply:
I think I missed the actual bug here :) 
I'd add a "Shutdown profiler" (or similar) line to the log _now_ as waiting for jitted/inlines is rather implicit (also won't work at all with your eager mode)

Reply:
I went for the lock check. I hope that the way I chose to check for that works also on network shares. We should test this!